tests/user_test.py::test_course_list assertion failure -- user dict last does not compare

See, for example, https://github.com/Plant-Tracer/webapp/actions/runs/6524598399

This test passed before the merge of PR 125 to main, but is failing after the merge.

```
=================================== FAILURES ===================================
_______________________________ test_course_list _______________________________

new_user = ('simsong+94081a@gmail.com', '96c69a50f7104ff687862e3af21b6442')

    def test_course_list(new_user):
        (user_email, api_key) = new_user
        user_dict = db.validate_api_key(api_key)
        user_id = user_dict['user_id']
        primary_course_id = user_dict['primary_course_id']
    
        recs1 = db.list_users(user_id=user_id)
    
        matches = [rec for rec in recs1 if rec['user_id']==user_id]
        assert(len(matches)>0)
    
        # Make sure that there is an admin in the course who is not the user
        recs2 = db.list_admins()
        matches = [rec for rec in recs2 if rec['course_id']==primary_course_id and rec['user_id']!=user_id]
        assert len(matches)==1
    
        # Make sure that the endpoint works
        with boddle(params={'api_key': api_key}):
            res = bottle_app.api_list_users()
        assert res['error'] is False
        assert len(recs1)==2
        assert len(res['result'])==2
>       assert res['result'] == recs1
E       AssertionError: assert [OrderedDict(...ast', None)])] == [OrderedDict(...ast', None)])]
E         At index 0 diff: OrderedDict([('name', 'Test User Name'), ('email', 'simsong+94081a@gmail.com'), ('primary_course_id', 2230), ('user_id', 2333), ('first', 1697379454), ('last', 1697379455)]) != OrderedDict([('name', 'Test User Name'), ('email', 'simsong+94081a@gmail.com'), ('primary_course_id', 2230), ('user_id', 2333), ('first', 1697379454), ('last', 1697379454)])
E         Full diff:
E           [
E            OrderedDict([('name', 'Test User Name'),
E                         ('email', 'simsong+94081a@gmail.com'),
E                         ('primary_course_id', 2230),
E                         ('user_id', 2333),
E                         ('first', 1697379454),
E         -               ('last', 1697379454)]),
E         ?                                 ^
E         +               ('last', 1697379455)]),
E         ?                                 ^
E            OrderedDict([('name', 'Test User Name'),
E                         ('email', 'simsong+admin+e518@gmail.com'),
E                         ('primary_course_id', 2230),
E                         ('user_id', 2332),
E                         ('first', None),
E                         ('last', None)]),
E           ]

tests/user_test.py:258: AssertionError
=============================== warnings summary ===============================
../../../.local/lib/python3.11/site-packages/bottle.py:38
  /home/runner/.local/lib/python3.11/site-packages/bottle.py:38: DeprecationWarning: 'cgi' is deprecated and slated for removal in Python 3.13
    import base64, cgi, email.utils, functools, hmac, itertools, mimetypes,\

tests/app_test.py: 11 warnings
  /home/runner/.local/lib/python3.11/site-packages/bottle.py:3383: DeprecationWarning: Absolute template path names are deprecated.
    fname = self.search(name, self.lookup)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html

---------- coverage: platform linux, python 3.11.5-final-0 -----------
Coverage XML written to file coverage.xml

=========================== short test summary info ============================
FAILED tests/user_test.py::test_course_list - AssertionError: assert [OrderedDict(...ast', None)])] == [OrderedDict(...ast', None)])]
  At index 0 diff: OrderedDict([('name', 'Test User Name'), ('email', 'simsong+94081a@gmail.com'), ('primary_course_id', 2230), ('user_id', 2333), ('first', 1697379454), ('last', 1697379455)]) != OrderedDict([('name', 'Test User Name'), ('email', 'simsong+94081a@gmail.com'), ('primary_course_id', 2230), ('user_id', 2333), ('first', 1697379454), ('last', 1697379454)])
  Full diff:
    [
     OrderedDict([('name', 'Test User Name'),
                  ('email', 'simsong+94081a@gmail.com'),
                  ('primary_course_id', 2230),
                  ('user_id', 2333),
                  ('first', 1697379454),
  -               ('last', 1697379454)]),
  ?                                 ^
  +               ('last', 1697379455)]),
  ?                                 ^
     OrderedDict([('name', 'Test User Name'),
                  ('email', 'simsong+admin+e518@gmail.com'),
                  ('primary_course_id', 2230),
                  ('user_id', 2332),
                  ('first', None),
                  ('last', None)]),
    ]
======= 1 failed, 24 passed, 1 skipped, 12 warnings in 112.41s (0:01:52) =======
make: *** [Makefile:48: coverage] Error 1
```

--- a/Makefile
+++ b/Makefile
@@ -28,12 +28,19 @@ flake8:
 	flake8 $(PYLINT_FILES)
 
 pytest:
+	make launch-local-mail
 	make touch
 	$(PYTHON) -m pytest . -v --log-cli-level=INFO
 
 pytest-debug:
+	make launch-local-mail
 	make touch
 	$(PYTHON) -m pytest . -v --log-cli-level=DEBUG
+
+pytest-quiet:
+	make launch-local-mail
+	make touch
+	$(PYTHON) -m pytest . --log-cli-level=ERROR
 
 create_localdb:
 	$(PYTHON) dbmaint.py --rootconfig etc/github_actions_mysql_rootconfig.ini --createdb actions_test --writeconfig etc/actions_test.ini

--- a/bottle_app.py
+++ b/bottle_app.py
@@ -131,7 +131,7 @@ def get_user_dict():
     userdict = db.validate_api_key(api_key)
     if not userdict:
         logging.warning("api_key %s is invalid",api_key)
-        raise bottle.HTTPResponse(body='', status=303, headers={ 'Location': '/'})
+        raise bottle.HTTPResponse(body='', status=303, headers={ 'Location': '/error'})
     return userdict
 
 def get_user_id():
@@ -198,6 +198,14 @@ GET_POST = [GET,POST]
 def func_root():
     """/ - serve the home page"""
     return page_dict()
+
+@bottle.route('/error', method=GET_POST)
+@view('error.html')
+def func_error():
+    auth.clear_cookie()
+    return {}
+
+
 
 @bottle.route('/about', method=GET_POST)
 @view('about.html')
@@ -310,11 +318,6 @@ def api_get_logs():
 
     logs    = db.get_logs(user_id=get_user_id(),**kwargs)
     return {'error':False, 'logs': logs}
-
-@bottle.route('/api/list-users', method=['POST'])
-def api_list_users():
-    """Get a list of the uses in the current course or all courses, depending on the user permissions."""
-    return db.list_users( user_id=get_user_id());
 
 @bottle.route('/api/register', method=['GET', 'POST'])
 def api_register():
@@ -522,7 +525,7 @@ def api_set_metadata():
 ##
 @bottle.route('/api/list-users', method=['POST'])
 def api_list_users():
-    return {'error': False, 'result': db.list_users(user_id=get_user_id())}
+    return {**{'error': False}, **db.list_users(user_id=get_user_id())}
 
 
 ##

--- a/db.py
+++ b/db.py
@@ -341,21 +341,29 @@ def delete_api_key(api_key):
 
 @log
 def list_users(*, user_id):
-    """Returns a dictionary of all the courses to which the user has access, and all of the people in them.
+    """Returns a dictionary with keys:
+    'users' - all the courses to which the user has access, and all of the people in them.
+    'courses' - all of the courses
     :param: user_id - the user doing the listing (determines what they can see)
     """
+    ret = {}
     cmd = """SELECT users.name AS name,users.email AS email,users.primary_course_id as primary_course_id, users.id AS user_id,
                     k.first as first,k.last as last
               FROM users LEFT JOIN
                       (select user_id,min(first_used_at) as first,max(last_used_at) as last from api_keys group by user_id) k
-              ON users.id=k.user_id
+                         ON users.id=k.user_id
               WHERE users.id=%s
                 OR users.primary_course_id IN (select primary_course_id from users where id=%s)
                 OR users.primary_course_id IN (select course_id from admins where user_id=%s)
                 OR %s IN (select user_id from admins where course_id=%s)
               ORDER BY primary_course_id,name,email"""
     args = (user_id, user_id,user_id,user_id,SUPER_ADMIN_COURSE_ID)
-    return dbfile.DBMySQL.csfr(get_dbreader(),cmd,args,asDicts=True)
+    ret['users'] = dbfile.DBMySQL.csfr(get_dbreader(),cmd,args,asDicts=True)
+
+    cmd = """SELECT id as course_id,course_name,course_section,max_enrollment from courses"""
+    args = []
+    ret['courses'] = dbfile.DBMySQL.csfr(get_dbreader(),cmd,args,asDicts=True)
+    return ret
 
 def list_admins():
     """Returns a list of all the admins"""

--- a/db.py
+++ b/db.py
@@ -218,12 +218,12 @@ def lookup_user(*, email=None, user_id=None, get_admin=None, get_courses=None):
 
     if get_admin:
         ret['admin'] = dbfile.DBMySQL.csfr(get_dbreader(),
-                                           """SELECT * from admin where user_id = %s""",
+                                           """SELECT * from admins where user_id = %s""",
                                            (user_id,), asDicts=True)
     if get_courses:
         ret['courses'] = dbfile.DBMySQL.csfr(get_dbreader(),
                                            """SELECT *,id as course_id from courses where id = %s
-                                           OR id in (select course_id from admin where user_id=%s)
+                                           OR id in (select course_id from admins where user_id=%s)
                                            """, (user_id,user_id),asDicts=True)
     return ret
 

--- a/static/planttracer.js
+++ b/static/planttracer.js
@@ -227,7 +227,9 @@ function row_checkbox_clicked( e ) {
 function row_pencil_clicked( e ) {
     console.log('row_pencil_clicked e=',e);
     const target = e.getAttribute('x-target-id'); // name of the target
+    console.log('target=',target);
     t = document.getElementById(target);       // element of the target
+    console.log('t=',t);
     const user_id  = t.getAttribute('x-user_id'); // property we are changing
     const movie_id = t.getAttribute('x-movie_id'); // property we are changing
     const property = t.getAttribute('x-property'); // property we are changing
@@ -459,22 +461,22 @@ function build_audit_table() {
         .then((data) => {
             if (data['error']!=false){
                 $('#message').html('error: '+data['message']);
-            } else {
-                console.log("data=",data);
-                let logs = data['logs'];
-                console.log("logs=",logs);
-                // get the columns
-                var columns = [];
-                for (const key in logs[0]) {
-                    //console.log(`${key}: ${logs{key}}`);
-                    columns.push( {'data':key, 'title':key } );
-                }
-                // make the data displayable
-                $('#audit').DataTable( {
-                    columns: columns,
-                    data: logs
-                });
+                return
             }
+            console.log("data=",data);
+            let logs = data['logs'];
+            console.log("logs=",logs);
+            // get the columns
+            var columns = [];
+            for (const key in logs[0]) {
+                //console.log(`${key}: ${logs{key}}`);
+                columns.push( {'data':key, 'title':key } );
+            }
+            // make the data displayable
+            $('#audit').DataTable( {
+                columns: columns,
+                data: logs
+            });
         });
 }
 
@@ -482,9 +484,47 @@ function build_audit_table() {
 ////////////////////////////////////////////////////////////////
 /// page: /users
 
+
+
+function list_users_data( users, course_array ) {
+    let current_course = null;
+    let div = $('#your-users');
+    let h = '<table>';
+    h += '<tbody>';
+    function user_html(user) {
+        var d1 = user.first ? new Date(user.first * 1000).toString() : "n/a";
+        var d2 = user.last ? new Date(user.last  * 1000).toString() : "n/a";
+        var ret = '';
+        if (current_course != user.primary_course_id) {
+            ret += `<tr><td colspan='4'>&nbsp;</td></tr>\n`;
+            ret += `<tr><th colspan='4'>Primary course: ${course_array[user.primary_course_id].course_name} (${user.primary_course_id})</th></tr>\n`;
+            ret += '<tr><th>Name</th><th>Email</th><th>First Seen</th><th>Last Seen</th></tr>\n';
+            current_course = user.primary_course_id;
+        }
+        ret +=  `<tr><td>${user.name} (${user.user_id}) </td><td>${user.email}</td><td>${d1}</td><td>${d2}</td></tr>\n`;
+        return ret;
+    };
+    users.forEach( user => ( h+= user_html(user) ));
+    h += '</tbody>';
+    div.html(h);
+}
+
 function list_users()
 {
     $('#message').html('Listing users...');
+    let formData = new FormData();
+    formData.append("api_key",  api_key); // on the upload form
+    fetch('/api/list-users', { method:"POST", body:formData })
+        .then((response) => response.json())
+        .then((data) => {
+            if (data['error']!=false){
+                $('#message').html('error: '+data['message']);
+                return;
+            }
+            var course_array = [];
+            data['courses'].forEach( course => (course_array[course.course_id] = course ));
+            list_users_data( data['users'], course_array);
+        });
 }
 
 

--- a/templates/error.html
+++ b/templates/error.html
@@ -1,0 +1,12 @@
+{% extends 'base.html' %}
+
+{% block body %}
+<h2> Session expired - You have been logged out</h2>
+
+<p> Your session has expired and you have been logged out. </p>
+
+<ul>
+<li> <a href='/register'> Register for a new account</a> (you must have a <b>course key</b>).</li>
+<li> <a href='/resend'> Resend your account login link</a></li>
+</ul>
+{% endblock %}

--- a/templates/users.html
+++ b/templates/users.html
@@ -12,7 +12,7 @@ Below is a list of all users registered on this system for your course
 {{ primary_course_name }}.
 {% endif %}
 
-<div id='user-list-class'></div>
+<div id='your-users'></div>
 
 
 

--- a/tests/app_test.py
+++ b/tests/app_test.py
@@ -18,8 +18,7 @@ sys.path.append(dirname(dirname(abspath(__file__))))
 from paths import STATIC_DIR
 import bottle_app
 
-API_KEY = os.getenv('TEST_USER_APIKEY')
-
+from user_test import new_course,new_user,new_movie,API_KEY
 
 def test_version():
     # With templates, res is just a string
@@ -51,7 +50,9 @@ def validate_html(html):
         raise
     assert "404 Not Found" not in data
 
-def test_templates():
+def test_templates(new_user):
+    api_key = new_user[API_KEY]
+
     # Test templates without an API_KEY
     with boddle(params={}):
         # These work without an api_key
@@ -74,7 +75,7 @@ def test_templates():
 
 
     # Test templates to see if they work with an API key
-    with boddle(params={'api_key': API_KEY}):
+    with boddle(params={'api_key': api_key}):
         validate_html(bottle_app.func_root())
         validate_html(bottle_app.func_about())
         validate_html(bottle_app.func_audit())
@@ -88,7 +89,8 @@ def test_templates():
         validate_html(bottle_app.func_users())
         validate_html(bottle_app.func_ver())
 
-def test_check_api_key():
+def test_check_api_key(new_user):
+    api_key = new_user[API_KEY]
     # no parameter should generate error
     with boddle(params={}):
         r = bottle_app.api_check_api_key()
@@ -100,6 +102,6 @@ def test_check_api_key():
         assert r['error'] == True
 
     # valid key should generate no error
-    with boddle(params={'api_key': API_KEY}):
+    with boddle(params={'api_key': api_key}):
         r = bottle_app.api_check_api_key()
         assert r['error'] == False, f'API_KEY {API_KEY} should be valid'

--- a/tests/endpoint_test.py
+++ b/tests/endpoint_test.py
@@ -18,13 +18,15 @@ import base64
 import pytest
 import logging
 import subprocess
-
+import copy
 
 sys.path.append(dirname(dirname(abspath(__file__))))
 
+import db
+from user_test import new_course, new_user, api_key
 
 # TEST_ENDPOINT = os.environ['TEST_ENDPOINT']
-TEST_USER_APIKEY = os.environ['TEST_USER_APIKEY']
+# TEST_USER_APIKEY = os.environ['TEST_USER_APIKEY']
 FRAME_FILES = glob.glob(os.path.join(TEST_DIR, "data", "frame_*.jpg"))
 FRAME_RE = re.compile(r"frame_(\d+).jpg")
 MOVIE_FILE_NAME = os.path.join(TEST_DIR, "data", "2019-07-31 plantmovie.mov")
@@ -51,16 +53,16 @@ def http_endpoint():
         line = p.stderr.readline()
         logging.info('%s', line)
         assert want in line
-    http_endpoint = f'http://127.0.0.1:{port}'
+    http_endpoint_url = f'http://127.0.0.1:{port}'
     # Make sure it is working and retry up to 10 times
 
     s = requests.Session()
-    s.mount(http_endpoint+'/ver',
+    s.mount(http_endpoint_url+'/ver',
             HTTPAdapter(max_retries=Retry(total=10, backoff_factor=0.2)))
-    r = s.get(http_endpoint+'/ver')
+    r = s.get(http_endpoint_url+'/ver')
     logging.info("Server running. r=%s", r)
     assert r.ok
-    yield http_endpoint
+    yield http_endpoint_url
     p.terminate()
     try:
         p.wait(timeout=1.0)
@@ -69,7 +71,6 @@ def http_endpoint():
         logging.error('KILL')
         p.kill()
         p.wait()
-
 
 @pytest.mark.skipif(SKIP_ENDPOINT_TEST, reason='SKIP_ENDPOINT_TEST set')
 def test_ver(http_endpoint):
@@ -89,26 +90,28 @@ def test_add(http_endpoint):
 
 
 @pytest.mark.skipif(SKIP_ENDPOINT_TEST, reason='SKIP_ENDPOINT_TEST set')
-def test_api_key(http_endpoint):
+def test_api_key(http_endpoint, api_key):
     r = requests.post(http_endpoint+'/api/check-api_key',
-                      {'api_key': TEST_USER_APIKEY})
+                      {'api_key': api_key})
     if (r.status_code != 200):
         raise RuntimeError(f"r.status_code={r.status_code}  r.text={r.text}")
     assert r.status_code == 200
     assert r.json()['error'] == False
-    assert r.json()['userinfo']['name'] == 'Test User'
+    assert r.json()['userinfo']['name'] == 'Test User Name'
 
     r = requests.post(http_endpoint+'/api/check-api_key',
-                      {'api_key': TEST_USER_APIKEY+'invalid'})
+                      {'api_key': 'invalid'})
     assert r.status_code == 200
     assert r.json()['error'] == True
 
 
 @pytest.mark.skipif(SKIP_ENDPOINT_TEST, reason='SKIP_ENDPOINT_TEST set')
-def test_upload_movie_frame_by_frame(http_endpoint):
-    """This tests creating a movie and uploading three frames using the frame-by-frame upload using an already existing test user"""
+def test_upload_movie_frame_by_frame(http_endpoint, api_key):
+    """This tests creating a movie and uploading
+    three frames using the frame-by-frame upload using an already existing test user"""
+
     assert len(FRAME_FILES) > 0
-    post_data = {'api_key': TEST_USER_APIKEY, 'title': 'Test Title at ' +
+    post_data = {'api_key': api_key, 'title': 'Test Title at ' +
                  time.asctime(), 'description': 'Test Upload'}
     r = requests.post(http_endpoint+'/api/new-movie', post_data)
     res = r.json()
@@ -120,7 +123,7 @@ def test_upload_movie_frame_by_frame(http_endpoint):
         print("uploading frame ", frame_number)
         with open(frame_file, 'rb') as f:
             r = requests.post(http_endpoint+'/api/new-frame', {
-                'api_key': TEST_USER_APIKEY,
+                'api_key': api_key,
                 'movie_id': movie_id,
                 'frame_msec': frame_number*200,
                 'frame_base64_data': base64.b64encode(f.read())})
@@ -129,25 +132,25 @@ def test_upload_movie_frame_by_frame(http_endpoint):
                 raise RuntimeError(json.dumps(r.json(), indent=4))
 
     # Now delete the movie
-    r = requests.post(http_endpoint+'/api/delete-movie', {
-        'api_key': TEST_USER_APIKEY,
-        'movie_id': movie_id})
-
-    # Now delete the movie
-    post_data2 = {'api_key': TEST_USER_APIKEY,
-                  'movie_id': movie_id}
-    r = requests.post(http_endpoint+'/api/delete-movie', post_data2)
+    r = requests.post(http_endpoint+'/api/delete-movie',
+                      { 'api_key': api_key,
+                        'movie_id': movie_id
+                       })
     res = r.json()
     assert res['error'] == False
 
+    # Purge the movie (to clean up)
+    db.purge_movie(movie_id = movie_id)
 
 @pytest.mark.skip(reason='not working yet')
-def test_upload_movie_data(http_endpoint):
+def test_upload_movie_data(http_endpoint, api_key):
     """This tests creating a movie and uploading the entire thing using base64 encoding and the existing test user"""
     assert len(FRAME_FILES) > 0
     with open(MOVIE_FILE_NAME, 'rb') as f:
         movie_base64_data = base64.b64encode(f.read())
-        post_data = {'api_key': TEST_USER_APIKEY, 'title': 'Test Title at '+time.asctime(), 'description': 'Test Upload',
+        post_data = {'api_key': api_key,
+                     'title': 'Test Title at '+time.asctime(),
+                     'description': 'Test Upload',
                      'movie_base64_data': movie_base64_data}
     r = requests.post(http_endpoint+'/api/new-movie', post_data)
     res = r.json()
@@ -160,3 +163,6 @@ def test_upload_movie_data(http_endpoint):
     r = requests.post(http_endpoint+'/api/delete-movie', post_data2)
     res = r.json()
     assert res['error'] == False
+
+    # Purge the movie (to clean up)
+    db.purge_movie(movie_id = movie_id)

--- a/tests/user_test.py
+++ b/tests/user_test.py
@@ -146,6 +146,15 @@ def test_new_user(new_user):
     (email, api_key) = new_user
     logging.info("email=%s api_key=%s", email, api_key)
 
+    # Try looking up the user
+    ret1 = db.lookup_user(email=email)
+    ret2 = db.lookup_user(user_id=ret1['user_id'], get_admin=True)
+    ret3 = db.lookup_user(user_id=ret1['user_id'], get_courses=True)
+
+    assert 'admin' in ret2
+    assert 'courses' in ret3
+
+
 def test_movie_upload(new_movie):
     """Create a new user, upload the movie, delete the movie, and shut down"""
     (movie_id, movie_title, api_key) = new_movie

--- a/tests/user_test.py
+++ b/tests/user_test.py
@@ -12,9 +12,9 @@ import uuid
 import base64
 import time
 import bottle
+import copy
 
 from os.path import abspath, dirname
-
 
 sys.path.append(dirname(dirname(abspath(__file__))))
 
@@ -33,84 +33,92 @@ TEST_ADMIN_NAME = 'Test User Name'
 
 MOVIE_FILENAME = os.path.join(MYDIR, "data", "2019-07-31 plantmovie.mov")
 
-def movie_list(api_key):
-    """Return a list of the movies"""
-    with boddle(params={"api_key": api_key}):
-        res = bottle_app.api_list_movies()
-    assert res['error'] == False
-    return res['movies']
-
-def get_movie(api_key, movie_id):
-    """Used for testing. Just pull the specific movie"""
-    movies = movie_list(api_key)
-    for movie in movies:
-        return movie
-
-    user_id = db.validate_api_key(api_key)['user_id']
-    logging.error("api_key=%s movie_id=%s user_id=%s",
-                  api_key, movie_id, user_id)
-    logging.error("len(movies)=%s", len(movies))
-    for movie in movies:
-        logging.error("%s", str(movie))
-    dbreader = db.get_dbreader()
-    logging.error("Full database: (dbreader: %s)", dbreader)
-    for movie in dbfile.DBMySQL.csfr(dbreader, "select * from movies", (), asDicts=True):
-        logging.error("%s", str(movie))
-    raise RuntimeError(f"No movie has movie_id {movie_id}")
-
-
+# keys for scaffolding dictionary
+ADMIN_EMAIL = 'admin_email'
+ADMIN_ID = 'admin_id'
+API_KEY  = 'api_key'
+COURSE_KEY = 'course_key'
+COURSE_NAME = 'course_name'
+USER_EMAIL = 'user_email'
+USER_ID    = 'user_id'
 
 ################################################################
 
 @pytest.fixture
 def new_course():
-    """Fixture to create a new course and then delete it. New course creates a new course admin and a new user for it"""
+    """Fixture to create a new course and then delete it.
+    New course creates a new course admin and a new user for it"""
+
     course_key = str(uuid.uuid4())[0:4]
     admin_email = TEST_ADMIN_EMAIL.replace('@', '+'+str(uuid.uuid4())[0:4]+'@')
+    course_name = course_key + "course name"
 
-    ct = db.create_course(course_key=course_key, course_name=course_key + "course name", max_enrollment=MAX_ENROLLMENT)['course_id']
+    ct = db.create_course(course_key=course_key,
+                          course_name=course_name,
+                          max_enrollment=MAX_ENROLLMENT)['course_id']
+
     admin_id = db.register_email(email=admin_email, course_key=course_key, name=TEST_ADMIN_NAME)['user_id']
+
     logging.info("generated course_key=%s  admin_email=%s admin_id=%s",course_key,admin_email,admin_id)
     db.make_course_admin(email=admin_email, course_key=course_key)
-    yield (course_key,admin_email)
+
+    yield {COURSE_KEY:course_key,
+           COURSE_NAME:course_name,
+           ADMIN_EMAIL:admin_email,
+           ADMIN_ID:admin_id
+           }
     db.remove_course_admin(email=admin_email, course_key=course_key)
     db.delete_user(email=admin_email)
     ct = db.delete_course(course_key=course_key)
     assert ct == 1                # returns number of courses deleted
-
 
 @pytest.fixture
 def new_user(new_course):
     """Creates a new course and a new user and yields (USER_EMAIL, api_key)
     Then deletes them. The course gets deleted by the new_course fixture.
     """
-    (course_key, admin_email) = new_course
+    cfg = copy.copy(new_course)
 
     user_email = TEST_USER_EMAIL.replace('@', '+'+str(uuid.uuid4())[0:6]+'@')
-    user_id = db.register_email(email=user_email, course_key=course_key, name=TEST_USER_NAME)['user_id']
+    user_id = db.register_email(email=user_email,
+                                course_key=cfg[COURSE_KEY],
+                                name=TEST_USER_NAME)['user_id']
     logging.info("generated user_email=%s user_id=%s",user_email, user_id)
 
     api_key = db.make_new_api_key(email=user_email)
     assert len(api_key) > 8
-    yield (user_email, api_key)
+
+    cfg[API_KEY] = api_key
+    cfg[USER_EMAIL] = user_email
+    cfg[USER_ID] = user_id
+
+    yield cfg
     ct = db.delete_api_key(api_key)
     assert ct == 1
     db.delete_user(email=user_email)
+
+@pytest.fixture
+def api_key(new_user):
+    """Simple fixture that just returns a valid api_key"""
+    yield new_user[API_KEY]
 
 
 @pytest.fixture
 def new_movie(new_user):
     """Create a new movie_id and return it"""
-    (email, api_key) = new_user
+    cfg = copy.copy(new_user)
 
-    MOVIE_TITLE = 'test movie title ' + str(uuid.uuid4())
+    api_key = cfg[API_KEY]
+    api_key_invalid = api_key+"invalid"
+
+    movie_title = 'test movie title ' + str(uuid.uuid4())
 
     with open(MOVIE_FILENAME, "rb") as f:
         movie_base64_data = base64.b64encode(f.read())
 
    # Try to uplaod the movie with an invalid key
-    with boddle(params={"api_key": api_key+'invalid',
-                        "title": MOVIE_TITLE,
+    with boddle(params={"api_key": api_key_invalid,
+                        "title": movie_title,
                         "description": "test movie description",
                         "movie_base64_data": movie_base64_data}):
         bottle_app.expand_memfile_max()
@@ -119,14 +127,18 @@ def new_movie(new_user):
 
     # Try to uplaod the movie all at once
     with boddle(params={"api_key": api_key,
-                        "title": MOVIE_TITLE,
+                        "title": movie_title,
                         "description": "test movie description",
                         "movie_base64_data": movie_base64_data}):
         res = bottle_app.api_new_movie()
     assert res['error'] == False
     movie_id = res['movie_id']
     assert movie_id > 0
-    yield (movie_id, MOVIE_TITLE, api_key)
+
+    cfg[MOVIE_ID] = movie_id
+    cfg[MOVIE_TITLE] = movie_title
+
+    yield cfg
 
     # Delete the movie we uploaded
     with boddle(params={'api_key': api_key,
@@ -138,16 +150,27 @@ def new_movie(new_user):
     db.purge_movie(movie_id=movie_id)
 
 
+################################################################
+## fixutre tests
+################################################################
+
+
+
+
 def test_new_course(new_course):
-    (course_key, admin_email) = new_course
+    cfg = copy.copy(new_course)
+    course_key = cfg[COURSE_KEY]
+    admin_email = cfg[ADMIN_EMAIL]
     logging.info("Created course %s", course_key)
 
 def test_new_user(new_user):
-    (email, api_key) = new_user
-    logging.info("email=%s api_key=%s", email, api_key)
+    cfg = copy.copy(new_course)
+    user_email = cfg[USER_EMAIL]
+    api_key = cfg[API_KEY]
+    logging.info("user_email=%s api_key=%s", email, api_key)
 
     # Try looking up the user
-    ret1 = db.lookup_user(email=email)
+    ret1 = db.lookup_user(email=user_email)
     ret2 = db.lookup_user(user_id=ret1['user_id'], get_admin=True)
     ret3 = db.lookup_user(user_id=ret1['user_id'], get_courses=True)
 
@@ -157,7 +180,10 @@ def test_new_user(new_user):
 
 def test_movie_upload(new_movie):
     """Create a new user, upload the movie, delete the movie, and shut down"""
-    (movie_id, movie_title, api_key) = new_movie
+    cfg = copy.copy(new_movie)
+    movie_id = cfg[MOVIE_ID]
+    movie_title = cfg[MOVIE_TITLE]
+    api_key = cfg[API_KEY]
 
     # Did the movie appear in the list?
     movies = movie_list(api_key)
@@ -173,8 +199,10 @@ def test_movie_upload(new_movie):
 
 def test_movie_update_metadata(new_movie):
     """try updating the metadata, and making sure some updates fail."""
-
-    (movie_id, movie_title, api_key) = new_movie
+    cfg = copy.copy(new_movie)
+    movie_id = cfg[MOVIE_ID]
+    movie_title = cfg[MOVIE_TITLE]
+    api_key = cfg[API_KEY]
 
     # Validate the old title
     assert get_movie(api_key, movie_id)['title'] == movie_title
@@ -229,6 +257,37 @@ def test_movie_update_metadata(new_movie):
 
     # Try to publish the movie with the course admin's API key. This should work
 
+################################################################
+## support functions
+################################################################
+
+
+def movie_list(api_key):
+    """Return a list of the movies"""
+    with boddle(params={"api_key": api_key}):
+        res = bottle_app.api_list_movies()
+    assert res['error'] == False
+    return res['movies']
+
+def get_movie(api_key, movie_id):
+    """Used for testing. Just pull the specific movie"""
+    movies = movie_list(api_key)
+    for movie in movies:
+        return movie
+
+    user_id = db.validate_api_key(api_key)['user_id']
+    logging.error("api_key=%s movie_id=%s user_id=%s",
+                  api_key, movie_id, user_id)
+    logging.error("len(movies)=%s", len(movies))
+    for movie in movies:
+        logging.error("%s", str(movie))
+    dbreader = db.get_dbreader()
+    logging.error("Full database: (dbreader: %s)", dbreader)
+    for movie in dbfile.DBMySQL.csfr(dbreader, "select * from movies", (), asDicts=True):
+        logging.error("%s", str(movie))
+    raise RuntimeError(f"No movie has movie_id {movie_id}")
+
+
 def test_get_logs():
     """Incrementally test each part of the get_logs functions"""
     dbreader = db.get_dbreader()
@@ -243,9 +302,12 @@ def test_get_logs():
         db.get_logs( user_id=0, ipaddr = "", security=security)
 
 def test_course_list(new_user):
-    (user_email, api_key) = new_user
+    cfg        = copy.copy(new_movie)
+    user_email = cfg[USER_EMAIL]
+    api_key    = cfg[API_KEY]
+
     user_dict = db.validate_api_key(api_key)
-    user_id = user_dict['user_id']
+    user_id   = user_dict['user_id']
     primary_course_id = user_dict['primary_course_id']
 
     recs1 = db.list_users(user_id=user_id)
@@ -271,8 +333,11 @@ def test_course_list(new_user):
 
 def test_log_search_user(new_user):
     """Currently we just run logfile queries and count the number of results."""
-    (user_email, api_key) = new_user
-    user_id = db.validate_api_key(api_key)['user_id']
+    cfg        = copy.copy(new_movie)
+    user_email = cfg[USER_EMAIL]
+    api_key    = cfg[API_KEY]
+
+    user_id  = db.validate_api_key(api_key)['user_id']
     dbreader = db.get_dbreader()
 
     ret = db.get_logs( user_id=user_id )
@@ -299,7 +364,9 @@ def test_log_search_user(new_user):
     assert(len(db.get_logs( user_id=user_id, course_key = course_key)) > 0)
 
     # Test to make sure that the course admin gets access to this user
-    admin_id = dbfile.DBMySQL.csfr(dbreader, "select user_id from admins where course_id=%s LIMIT 1", (course_id,))[0]
+    admin_id = dbfile.DBMySQL.csfr(dbreader,
+                                   "SELECT user_id FROM admins WHERE course_id=%s LIMIT 1",
+                                   (course_id,))[0]
     assert(len(db.get_logs( user_id=admin_id, log_user_id=user_id, course_id = course_id)) > 0)
     assert(len(db.get_logs( user_id=admin_id, log_user_id=user_id, course_key = course_key)) > 0)
 
@@ -307,7 +374,11 @@ def test_log_search_user(new_user):
     assert(len(db.get_logs( user_id=user_id, ipaddr="0.0.0.0"))==0)
 
 def test_log_search_movie(new_movie):
-    (movie_id, movie_title, api_key) = new_movie
+    cfg        = copy.copy(new_movie)
+    api_key    = cfg[API_KEY]
+    movie_id   = cfg[MOVIE_ID]
+    movie_title= cfg[MOVIE_TITLE]
+
     dbreader = db.get_dbreader()
     res = dbfile.DBMySQL.csfr(dbreader, "select user_id from movies where id=%s", (movie_id,))
     user_id = res[0][0]

--- a/tests/user_test.py
+++ b/tests/user_test.py
@@ -41,6 +41,8 @@ COURSE_KEY = 'course_key'
 COURSE_NAME = 'course_name'
 USER_EMAIL = 'user_email'
 USER_ID    = 'user_id'
+MOVIE_ID = 'movie_id'
+MOVIE_TITLE = 'movie_title'
 
 ################################################################
 
@@ -164,10 +166,10 @@ def test_new_course(new_course):
     logging.info("Created course %s", course_key)
 
 def test_new_user(new_user):
-    cfg = copy.copy(new_course)
+    cfg = copy.copy(new_user)
     user_email = cfg[USER_EMAIL]
     api_key = cfg[API_KEY]
-    logging.info("user_email=%s api_key=%s", email, api_key)
+    logging.info("user_email=%s api_key=%s", user_email, api_key)
 
     # Try looking up the user
     ret1 = db.lookup_user(email=user_email)
@@ -302,7 +304,7 @@ def test_get_logs():
         db.get_logs( user_id=0, ipaddr = "", security=security)
 
 def test_course_list(new_user):
-    cfg        = copy.copy(new_movie)
+    cfg        = copy.copy(new_user)
     user_email = cfg[USER_EMAIL]
     api_key    = cfg[API_KEY]
 
@@ -333,7 +335,7 @@ def test_course_list(new_user):
 
 def test_log_search_user(new_user):
     """Currently we just run logfile queries and count the number of results."""
-    cfg        = copy.copy(new_movie)
+    cfg        = copy.copy(new_user)
     user_email = cfg[USER_EMAIL]
     api_key    = cfg[API_KEY]
 

--- a/tests/user_test.py
+++ b/tests/user_test.py
@@ -249,8 +249,9 @@ def test_course_list(new_user):
     primary_course_id = user_dict['primary_course_id']
 
     recs1 = db.list_users(user_id=user_id)
+    users1 = recs1['users']
 
-    matches = [rec for rec in recs1 if rec['user_id']==user_id]
+    matches = [user for user in users1 if user['user_id']==user_id]
     assert(len(matches)>0)
 
     # Make sure that there is an admin in the course who is not the user
@@ -262,9 +263,10 @@ def test_course_list(new_user):
     with boddle(params={'api_key': api_key}):
         res = bottle_app.api_list_users()
     assert res['error'] is False
-    assert len(recs1)==2
-    assert len(res['result'])==2
-    assert res['result'] == recs1
+    users2 = res['users']
+    assert len(users2)==2
+    assert users1[0]['name'] == users2[0]['name']
+    assert users1[0]['email'] == users2[0]['email']
 
 
 def test_log_search_user(new_user):

